torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -955,133 +962,341 @@ def _run_param_init_fn():
+                # memory efficiency.
+                for param in orig_params:
+                    _free_storage(param)
+                    # Synchronize to free the memory immediately
+                    torch.cuda.current_stream().synchronize()
+            self.params.append(new_handle.flat_param)
+        for handle in self._handles:
+            self._init_param_attributes(handle.flat_param)
+            handle._mode = HandleMode.UNINITIALIZED
+        self._register_flat_params()
+        # Re-register the forward hooks since the handle construction changed
+        self._register_pre_forward_hooks()
+        self._register_post_forward_hooks()
+
+    def _check_single_device_module(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+    ) -> None:
+        """Raises an error if ``module`` has parameters on multiple devices
+        (ignoring the parameters in ``ignored_params``)."""
+        devices = set(
+            param.device for param in module.parameters()
+            if param not in ignored_params and not isinstance(param, FlatParameter)
+        )
+        if len(devices) > 1:
+            raise RuntimeError(
+                f"FSDP only supports single device modules, but got params on {devices}"
+            )
+
+    def _check_param_device_with_device_id(
+        self,
+        param: nn.Parameter,
+        device_id: torch.device,
+    ):
+        """Raises an error if ``param`` is on a non-CPU device that does not
+        match ``device_id``."""
+        if param.device != torch.device("cpu") and param.device != device_id:
+            raise RuntimeError(
+                f"Module on rank {self.rank} is given device_id argument "
+                f"{self.device_id}, but is on {param.device}. "
+                " Either move module before FSDP init or omit device_id argument."
+            )
+
+    def _warn_cpu_init(self):
+        warnings.warn(
+            "Module is put on CPU and will thus have flattening and sharding"
+            " run on CPU, which is less efficient than on GPU. We recommend passing in "
+            "`device_id` argument which will enable FSDP to put module on GPU device,"
+            " module must also be on GPU device to work with `sync_module_states=True` flag"
+            " which requires GPU communication."
+        )
+
+    def _move_module_to_device_id(self, module: nn.Module) -> None:
+        """
+        Moves ``module`` appropriately depending on ``self.device_id`` and
+        whether ``module`` is on CPU.
+        
+        - If a ``device_id`` was specified to the constructor, then this moves
+        ``module`` to the corresponding device. This should be done before
+        setting ``self.compute_device`` to ensure they align.
+        - If no ``device_id`` was specified to the constructor, then we warn the
+        user, asking for a ``device_id`` for faster FSDP initialization.
+        """
+        param_gen = module.parameters()
```

Comment:
To-do: fix this e.g. with a subroutine that returns `None` if no original parameters in `module.parameters()` and the first original parameter otherwise.